browser/ui/commander/commander_service.cc

```diff
@@ -0,0 +1,245 @@
+    return;
+  }
+
+  // Increment the current result set id - we don't want any commands from this
+  // set to be reused after we've selected a command.
+  // Note: This needs to happen before beginning a composite command, to ensure
+  // that the generated model uses right right |result_set_id|.
+  current_result_set_id_++;
+
+  auto* item = items_[command_index].get();
+  if (item->GetType() == CommandItem::Type::kOneShot) {
+    std::move(absl::get<base::OnceClosure>(item->command)).Run();
+    Hide();
+  } else {
+    auto composite_command =
+        absl::get<CommandItem::CompositeCommand>(item->command);
+    std::tie(prompt_, composite_command_provider_) = composite_command;
+    Show();
+  }
+}
+
+std::vector<CommandItemModel> CommanderService::GetItems() {
+  std::vector<CommandItemModel> result;
+  base::ranges::transform(items_, std::back_inserter(result), FromCommand);
+  return result;
+}
+
+int CommanderService::GetResultSetId() {
+  return current_result_set_id_;
+}
+
+const std::u16string& CommanderService::GetPrompt() {
+  return prompt_;
+}
+
+void CommanderService::Toggle() {
+  if (IsShowing()) {
+    Hide();
+  } else {
+    Show();
+  }
+}
+
+void CommanderService::Show() {
+  // Note: This posts a task because we can't change the Omnibox text while
+  // autocompleting.
+  base::SequencedTaskRunner::GetCurrentDefault()->PostTask(
+      FROM_HERE,
+      base::BindOnce(
+          [](base::WeakPtr<CommanderService> service) {
+            if (auto* omnibox = service->GetOmnibox()) {
+              omnibox->SetFocus(true);
+
+              auto text = base::StrCat({commander::kCommandPrefix, u" "});
+              omnibox->SetUserText(text);
+              omnibox->SetCaretPos(text.size());
+              service->UpdateText(true);
+            }
+          },
+          weak_ptr_factory_.GetWeakPtr()));
+}
+
+void CommanderService::Hide() {
+  // Note: This posts a task because we can't change the Omnibox text while
+  // autocompleting.
+  base::SequencedTaskRunner::GetCurrentDefault()->PostTask(
+      FROM_HERE, base::BindOnce(
+                     [](base::WeakPtr<CommanderService> service) {
+                       service->Reset();
+
+                       if (auto* omnibox = service->GetOmnibox();
+                           omnibox && service->IsShowing()) {
+                         omnibox->RevertAll();
+                       }
+                     },
+                     weak_ptr_factory_.GetWeakPtr()));
+}
+
+void CommanderService::Reset() {
+  current_result_set_id_++;
+  items_.clear();
+  prompt_.clear();
+  last_searched_.clear();
+  last_browser_ = nullptr;
+  if (composite_command_provider_) {
+    composite_command_provider_.Reset();
+  }
+  NotifyObservers();
+}
+
+OmniboxView* CommanderService::GetOmnibox() const {
+  auto* browser = chrome::FindLastActiveWithProfile(profile_);
+  if (!browser) {
+    return nullptr;
+  }
+  return browser->window()->GetLocationBar()->GetOmniboxView();
```

Comment:
The only comment I see on `window` is this one:

```
  // |window()| will return NULL if called before |CreateBrowserWindow()|
  // is done.
```

but that doesn't mean it can't be. Maybe I'll add a DCHECK?